### ISSUE TITLE
LINKER ERROR when building Windows 10 Provisioning Script

### ISSUE THREAD

**Issue by HyugaTatsu** [osquery/osquery#6247]
Hi, I'm building my own extension following this link https://osquery.readthedocs.io/en/stable/development/building/ and https://github.com/trailofbits/osquery-extensions/blob/master/README.md
For some reason, I have to use the version 3.3.2
- What I have done is
    + set my project at C driver, so my path to osquery is: `C:\osquery\tmp\osquery`
    + switch git tag to version `3.3.2` for building my own extension
    + run command as Administrator
    + running `.\tools\make-win64-dev-env.bat`
- My problem is: when I ran `.\tools\make-win64-binaries.bat`, I met this LINKER error even when I did not add my own extension to folder `external`
```
 Creating library C:/osquery/tmp/osquery/build/windows10/osquery/Release/osqueryd.lib and object C:/osquery/tmp/osquery/build/windows10/osquery/Release/osqueryd.explibeay32.lib(b_sock.obj) : error LNK2001: unresolved external symbol __imp_htonl [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    osquery.lib(impl_thrift.obj) : error LNK2001: unresolved external symbol __imp_htonl [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    osquery.lib(rocksdb_database.obj) : error LNK2001: unresolved external symbol __imp_htonl [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    thriftmt.lib(TBufferTransports.obj) : error LNK2001: unresolved external symbol __imp_htonl [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    libeay32.lib(bss_conn.obj) : error LNK2001: unresolved external symbol __imp_htonl [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    osquery.lib(impl_thrift.obj) : error LNK2001: unresolved external symbol __imp_htons [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    osquery.lib(rocksdb_database.obj) : error LNK2001: unresolved external symbol __imp_htons [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    libeay32.lib(bss_conn.obj) : error LNK2001: unresolved external symbol __imp_htons [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
    ...
    C:\osquery\tmp\osquery\build\windows10\osquery\Release\osqueryd.exe : fatal error LNK1120: 110 unresolved externals [C:\osquery\tmp\osquery\build\windows10\osquery\daemon.vcxproj]
```
- I need suggestions for fixing these LINKER errors. Any suggestions would be great. Thank you very much
![image](https://user-images.githubusercontent.com/33058526/74701873-d1395e00-523a-11ea-821b-64abdec79af1.png)
- P.s: what I understand is to build osqueryd, I will need to build the `daemon` project. This leads to the fact that I have to build `libosquery` and `libosquery_additional` projects to get `osquery.lib` and `osquery_additional.lib`. The problem shows up when the msbuild try to link the libraries for osqueryd

**Comment by muffins** [comment]
Oof. Yeah attempting to build and link an extension on the 3.3.2 version of osquery might be rough. I think we should leverage this issue to focus on re-building/writing all of the docs on compiling windows extensions, as a lot has changed and this process isn't what it used to be. For starters, I'd recommend avoiding any of the build scripts we used to leverage, and instead start with building osqueryd normally via `cmake` as detailed in the readthedocs. This should at least produce a libosquery for you to then link against in a different visual studio project. I know that's pretty vague, but I need to revisit this whole flow, I hope that's at least a helpful first step.

I can't guarantee much on timeline, but I'll try and revisit building native Windows extensions sometime this week and see if I can get a write-up on how to do it.

**Comment by muffins** [comment]
Better yet, I know the ToB folks make heavy use of native extensions on Windows, @Smjert, @alessandrogario, or @mike-myers-tob might have much better insight than I do, and could possibly point you to a very quick and easy mechanism to build osquery native extensions for Windows.

**Comment by Smjert** [comment]
I think I've answered some of the questions on Slack and basically I made him use osquery 4.x because I do recall that when we were working on osql, based on 3.3.2, we couldn't even build it, so we had to do some changes.

That been said, we have this branch https://github.com/trailofbits/osquery-extensions/tree/extension_porting_4.0.1 that can be taken as an example to create extensions on 4.x and it can be compared to version on master.

Building should be easy, you just need to link the folder where your extension project exists under the externals folder in the osquery source, then configure again and build osquery, as it's written in the osquery building guide on the wiki.

What changes a little is how, through CMake, you let osquery know that an extension exist and has to be built; for that I suggest looking into https://github.com/osquery/osquery/blob/master/external/cmake/cmakelibs.cmake and comparing the use we do in our repo.
Then another difference is where some classes are, or header paths, because some actually moved in other folders, or some take advantage of the header namespace feature Buck introduces.

The last thing I recall is that the return type of the table gen* function changed from QueryData to TableRows, which adds some levels of indirections to actually support multiple return types.
This unfortunately is not a finished work, because I think the attempt was to incrementally have extensions (and then probably also internal tables?) using strictly typed structures as return types, instead of maps with string values.

So long story short, `Row` cannot be anymore used as it is, together with `QueryData`; the equivalent is `DynamicTableRowHolder` that can be push (or emplace)-moved inside a `TableRows` instance.

We definitely need to update several parts of the docs because it makes references to "the old ways", and we also have to choose how to deal with this new row type, because it definitely has some rough edges.

**Comment by HyugaTatsu** [comment]
Wow, definitely yes, I had changed to the lastest version of osquery: 4.2.0, and combine with new branch https://github.com/trailofbits/osquery-extensions/tree/extension_porting_4.0.1  of osquery, I had built tob extension successfully. Thank you all for lovely support. And I will close this topic to bury the old version :))
